### PR TITLE
unfurling: add settings module and dev mode conversation storage system CORE-9242

### PR DESCRIPTION
**Pull request by mmaxim**
Patch does the following:

1.) Introduces a new interface `ConversationBackedStorage` with two simple accessor methods with the purpose if being implemented by a chat conversation.
2.) Add `DevConversationBackedStorage` which writes `TEXT` messages into `DEV` conversations, and always uses the contents of the latest message in the return type. Data is JSON encoded into the chat text messages.
3.) Add `unfurl.Settings` which manages the unfurling settings via a `ConversationBackedStorage`.
4.) Add simple methods and new types for tracking settings.

### CODE REVIEW

File: go/chat/devstorage.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+			}),
+		}, 0, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *DevConversationBackedStorage) Get(ctx context.Context, name string, dest interface{}) (found bool, err error) {
+	defer s.Trace(ctx, func() error { return err }, "Get(%s)", name)()
+	if s.G().ActiveDevice == nil {
```
Comment (joshblum) [inline go/chat/devstorage.go]: can you move `assertLoggedInUID` from the server to `utils` so we can use it throughout the chat package? that checks `ActiveDevice.HaveKeys()` and we should keep it consistent (i don't think there is a practical difference but we do this check everywhere so it's confusing if it's different)

File: go/chat/devstorage.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		return false, libkb.LoggedInError{}
+	}
+	uid := gregor1.UID(s.G().ActiveDevice.UID().ToBytes())
+	username := s.G().ActiveDevice.Username(libkb.NewMetaContext(ctx, s.G().ExternalG())).String()
+
+	convs, err := FindConversations(ctx, s.G(), s.DebugLabeler, true, s.ri, uid, username,
```
Comment (joshblum) [inline go/chat/devstorage.go]: why not use `NewConversation` here as well (functions as get or create right?) and then both methods can just have a shared `getConversation` method

File: go/chat/unfurl/settings.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	var mr modeRecord
+	found, err := s.storage.Get(ctx, settingsModeName, &mr)
+	if err != nil {
+		return res, err
+	}
+	if !found {
+		s.Debug(ctx, "Get: no mode setting found, using whitelisted")
+		mr.Mode = chat1.UnfurlMode_WHITELISTED
+	}
```
Comment (joshblum) [inline go/chat/unfurl/settings.go]: abort whitelist fetch if mode isn't whitelisted?

File: go/chat/unfurl/settings.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	defer s.Trace(ctx, func() error { return err }, "WhitelistAdd(%s)", domain)()
+	var wr whitelistRecord
+	found, err := s.storage.Get(ctx, settingsWhitelistName, &wr)
+	if err != nil {
+		return err
+	}
+	if !found {
+		wr.Whitelist = nil
+	}
+	for _, w := range wr.Whitelist {
```
Comment (joshblum) [inline go/chat/unfurl/settings.go]: why not make this a map?

File: go/chat/devstorage.go (hunk 2 above)
Comment (mmaxim) [reply go/chat/devstorage.go]: Didn't want it to create it if it wasn't already there.

File: go/chat/unfurl/settings.go (hunk 3 above)
Comment (mmaxim) [reply go/chat/unfurl/settings.go]: I wanted to always get the whitelist for rendering the settings screen, where I think the whitelist will still show if you are in always or never mode.

File: go/chat/unfurl/settings.go (hunk 4 above)
Comment (mmaxim) [reply go/chat/unfurl/settings.go]: Yeah probably better, I'll change it.

File: go/chat/devstorage.go (hunk 1 above)
Comment (mmaxim) [reply go/chat/devstorage.go]: Sure

File: go/chat/unfurl/settings.go (hunk 4 above)
Comment (mmaxim) [reply go/chat/unfurl/settings.go]: Actually I like it as a list better. The most common operation will be `Get`, and if I change to a map I need to sort it every time (or have some param like should sort). I know the code is a little uglier, but its just contained in this one file.

**Comment by mmaxim** [comment]
@joshblum ok made `assertLoggedInUID` change, let me know if you accept my other replies.

File: go/chat/unfurl/settings.go (hunk 3 above)
Comment (joshblum) [reply go/chat/unfurl/settings.go]: ah makes sense

File: go/chat/unfurl/settings.go (hunk 4 above)
Comment (joshblum) [reply go/chat/unfurl/settings.go]: i feel it should only be sorted before it's rendered which is probably the least frequent op.  should there be some `IsValid(domain) bool` method to wrap calling `Get` and handing the mode/whitelist?

Comment (mmaxim) [reply go/chat/unfurl/settings.go]: Changed my mind again, I think map is the way.

Comment (joshblum) [reply go/chat/unfurl/settings.go]: heheok

**Comment by mmaxim** [comment]
Going to merge since I think I hit all the feedback, let me know if there is anythign else.

**Pull request merged**